lib/srv/termmanager.go

```diff
@@ -122,56 +135,43 @@ func (g *TermManager) Write(p []byte) (int, error) {
 }
 
 func (g *TermManager) Read(p []byte) (int, error) {
-	if len(g.remaining) > 0 {
-		n := copy(p, g.remaining)
-		g.remaining = g.remaining[n:]
-		return n, nil
-	}
-
-	q := make(chan struct{})
-	defer close(q)
-	c := make(chan bool)
-	go func() {
-		g.readStateUpdate.L.Lock()
-		defer g.readStateUpdate.L.Unlock()
-
-		for {
-			g.mu.Lock()
-			on := g.on
-			g.mu.Unlock()
-
-			select {
-			case c <- on:
-			case <-q:
-				close(c)
-				return
-			}
-
-			g.readStateUpdate.Wait()
-		}
-	}()
+	g.mu.Lock()
+	state := g.state
+	g.mu.Unlock()
 
-	on := <-c
 	for {
-		if !on {
-			select {
-			case <-g.closed:
-				return 0, io.EOF
-			case on = <-c:
-				continue
+		// wait until data flow is enabled and there is data to be read, or the session is terminated.
+		func() {
+			for state == dataFlowOff || len(g.remaining) == 0 {
```

Comment:
Do we need both the outer and inner loops? Do you think we could simplify?

It looks to me that this only ever loops twice if the result of `g.remaining = append(g.remaining, data...)` is still an empty slice, is that something that does happen?